osgi/karaf/features/src/main/resources/features.xml

```diff
@@ -216,8 +216,9 @@
         </capability>
     </feature>
     <feature name="cxf-http-undertow" version="${project.version}">
-        <feature version="${project.version}">cxf-http</feature>
+        <feature version="${project.version}">cxf-core</feature>
```

Comment:
@ffang would appreciate you input on this change. 

Had to remove the dependency on the `http` feature. It causes the issue with `servlet-api` dependencies (see please http://karaf.922171.n3.nabble.com/karaf-4-3-0-unable-to-install-hawtio-core-with-pax-http-undertow-td4059219.html).

[ERROR]                                                                                                                               
[ERROR] Chain 1:                                                                                                                      
[ERROR]   org.ops4j.pax.web.pax-web-undertow [org.ops4j.pax.web.pax-web-undertow/7.3.13]                                              
[ERROR]     import: (&(osgi.wiring.package=javax.servlet)(version>=3.1.0)(!(version>=5.0.0)))                                         
[ERROR]      |                                                                                                                        
[ERROR]     export: osgi.wiring.package: javax.servlet                                                                                
[ERROR]   javax.servlet-api [javax.servlet-api/3.1.0]                                                                                 
[ERROR]                                                                                                                               
[ERROR] Chain 2:                                                                                                                      
[ERROR]   org.ops4j.pax.web.pax-web-undertow [org.ops4j.pax.web.pax-web-undertow/7.3.13]                                              
[ERROR]     import: (&(osgi.wiring.package=io.undertow.servlet)(version>=2.0.0)(!(version>=2.1.0)))                                   
[ERROR]      |                                                                                                                        
[ERROR]     export: osgi.wiring.package=io.undertow.servlet; uses:=javax.servlet                                                      
[ERROR]   io.undertow.servlet [io.undertow.servlet/2.0.33.Final]                                                                      
[ERROR]     import: (&(osgi.wiring.package=javax.servlet)(version>=4.0.0)(!(version>=5.0.0)))                                         
[ERROR]      |                                                                                                                        
[ERROR]     export: osgi.wiring.package: javax.servlet                                                                                
[ERROR]   javax.servlet-api [javax.servlet-api/4.0.0]                                                                                 
[ERROR] Repositories: {                                                                                                               
[ERROR]         file:C:\Users\retab\Workspaces\apache-cxf-git\cxf-git\osgi\karaf\features\target/classes/features.xml                 
[ERROR]         mvn:org.apache.karaf.features/framework/4.3.1/xml/features                                                            
[ERROR]         mvn:org.apache.karaf.features/spring/4.3.1/xml/features                                                               
[ERROR]         mvn:org.apache.karaf.features/standard/4.3.1/xml/features                                                             
[ERROR]         mvn:org.ops4j.pax.cdi/pax-cdi-features/1.1.4/xml/features                                                             
[ERROR]         mvn:org.ops4j.pax.web/pax-web-features/7.3.13/xml/features                                                            
[ERROR] }                                                                                                                             

What do you think?

Reply:
Hi @reta ,

Thanks for bringing this up, and this part is tricky indeed. The problem from that some features in CXF now still require servlet api 3.x(like jetty), but others require servlet 4.x

To address this, we need to introduce a servlet-compatibility fragment bundle, which is like
```
<plugin>
                <groupId>org.apache.felix</groupId>
                <artifactId>maven-bundle-plugin</artifactId>
                <configuration>
                    <instructions>
                        <Bundle-SymbolicName>${project.groupId}.${project.artifactId}</Bundle-SymbolicName>
                        <Fragment-Host>javax.servlet-api</Fragment-Host>
                        <Export-Package>
                            javax.servlet;uses:="javax.servlet.annotation,javax.servlet.descriptor";version=3.1.0,
                            javax.servlet.annotation;uses:=javax.servlet;version=3.1.0,
                            javax.servlet.http;uses:=javax.servlet;version=3.1.0,
                            javax.servlet.descriptor;version=3.1.0
                        </Export-Package>
                    </instructions>
                </configuration>
            </plugin>
```
And in the org.apache.karaf.features.xml, we need to patch pax-http-undertow features with this fragment bundle like
```
   <featureReplacements>
        <replacement mode="merge">
            <feature name="pax-http-undertow" version="7.3.13">
                <f:bundle start-level="35">mvn:groupId/servlet-compatibility/version</f:bundle>
            </feature>
        </replacement>
    </featureReplacements>
```

This is the solution we used in FUSE, and we have this fragment bundle from Red Hat maven repo. We can reuse this from FUSE(need to add Red Hat maven repo in build), or we can add this fragment bundle in CXF(for now). Or something I think better, we should add this in the OPS4J PAX-WEB, to avoid patching pax-http-undertow feature downstream.

Best Regards
Freeman

Reply:
Thanks a mill @ffang , I have also tried slightly different approach with `featureReplacements` but did not succeed, the trick with the  fragment bundle is definitely a great one, refactoring the PR ....

Reply:
@ffang my apologies, mind please taking another look, refactored the change to use compatibility bundle and feature processing, thanks a lot!

Reply:
LGTM, thanks a lot!